android/src/main/kotlin/com.netguru.multiplatformstorage/AndroidStorageProvider.kt

```diff
@@ -25,19 +52,123 @@ internal object AndroidStorageProvider {
+            clearKeys(keyStore, preferences)
+            generateKeystoreKey(KEYSTORE_VALUES_KEY_ALIAS, keyStore)
+            generateKeystoreKey(KEYSTORE_VALUES_IV_ALIAS, keyStore)
+            generateKeystoreKey(KEYSTORE_KEYS_KEY_ALIAS, keyStore)
+            preferences.edit {
+                putString(VALUES_KEY_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_VALUES_KEY_ALIAS)))
+                putString(VALUES_IV_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_VALUES_IV_ALIAS)))
+                putString(KEYS_KEY_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_KEYS_KEY_ALIAS)))
+            }
+        }
+
+        val valuesKey = getDecryptedKey(VALUES_KEY_ALIAS, KEYSTORE_VALUES_KEY_ALIAS, preferences, keyStore)
+        val valuesIv = getDecryptedKey(VALUES_IV_ALIAS, KEYSTORE_VALUES_IV_ALIAS, preferences, keyStore)
+        val keysKey = getDecryptedKey(KEYS_KEY_ALIAS, KEYSTORE_KEYS_KEY_ALIAS, preferences, keyStore)
+
+
+        return BinaryPreferencesBuilder(appContext)
+            .migrateFrom(getSharedPreferences(name))
+            .keyEncryption(XorKeyEncryption(keysKey))
+            .valueEncryption(AesValueEncryption(valuesKey, valuesIv))
+            .name(name ?: "default")
+            .build()
+    }
+
+    private fun clearKeys(keyStore: KeyStore, sharedPreferences: SharedPreferences) {
+        with(keyStore) {
+            deleteEntry(KEYSTORE_VALUES_KEY_ALIAS)
+            deleteEntry(KEYSTORE_VALUES_IV_ALIAS)
+            deleteEntry(KEYSTORE_KEYS_KEY_ALIAS)
+        }
+        sharedPreferences.edit {
+            remove(VALUES_KEY_ALIAS)
+            remove(VALUES_IV_ALIAS)
+            remove(KEYS_KEY_ALIAS)
+        }
+    }
+
+    private fun keysExist(keyStore: KeyStore, preferences: SharedPreferences) =
+        keyStore.containsAlias(KEYSTORE_VALUES_KEY_ALIAS) && keyStore.containsAlias(KEYSTORE_KEYS_KEY_ALIAS)
+                && keyStore.containsAlias(KEYSTORE_VALUES_IV_ALIAS) && preferences.contains(VALUES_IV_ALIAS)
+                && preferences.contains(VALUES_KEY_ALIAS) && preferences.contains(KEYS_KEY_ALIAS)
 
-        val sharedPrefs = if (name == null) {
-            PreferenceManager.getDefaultSharedPreferences(appContext)
-        } else {
-            appContext.getSharedPreferences(name, Context.MODE_PRIVATE)
+    private fun generateKeystoreKey(keyAlias: String, keyStore: KeyStore) = with(keyStore) {
+        val aesKeyGenSpec =
+            KeyGenParameterSpec.Builder(keyAlias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
+                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
+                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+                .build()
+        with(KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_NAME)) {
+            init(aesKeyGenSpec)
+            generateKey()
         }
+    }
+
+    private fun generateEncryptedKey(key: Key) =
```

Comment:
This method does something slightly different than generation of encrypted key. Its encoding the encrypted key as Base64 string ;)